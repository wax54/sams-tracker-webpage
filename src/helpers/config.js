export const timeFrames = {
    day: {
        title: "Every Day",
        val: 1,
        roundPlaces: 2
    },
    week: {
        title: "Every Week",
        val: 7,
        roundPlaces: 1
    },
    month: {
        title: "Every Month",
        val: 30,
        roundPlaces: 0
    }
}