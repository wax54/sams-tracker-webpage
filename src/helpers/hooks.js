import { useState, useEffect } from 'react';

function getWindowDimensions() {
    const { innerWidth: width, innerHeight: height } = window;
    return {
        width,
        height
    };
}

export function useWindowDimensions() {
    const [windowDimensions, setWindowDimensions] = useState(getWindowDimensions());

    useEffect(() => {
        function handleResize() {
            setWindowDimensions(getWindowDimensions());
        }

        window.addEventListener('resize', handleResize);
        return () => window.removeEventListener('resize', handleResize);
    }, [setWindowDimensions]);

    return windowDimensions;
}


export function useFormFields(init={}) {
    const [formData, setFormData] = useState(init);
    
    const handleChange = evt => {
        const {name, value} = evt.target;
        setFormData(data => ({ ...data, [name]: value }));
    };
    const resetForm = () => setFormData(init);
    return [formData, handleChange, resetForm];
}