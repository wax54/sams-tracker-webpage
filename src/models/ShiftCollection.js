/**
 * A shift class, representing time spent at a certain date on a thing for a reason
 * 
 */
class Shift {
    /**
     * 
     * @param { Date } start 
     * @param { string } type 
     * @param { string } category  
     * @param { Date } stop
     * @param { Number } id
     * @param { Number } u_id
     */
    constructor({ start, type, category, stop, id, u_id }) {
        this.id = id;
        this.u_id = u_id;
        if (start instanceof Date) this.start = start;
        else this.start = new Date(start);

        this.type = type;
        this.category = category;
        if (stop) {
            if (stop instanceof Date) this.stop = stop;
            else this.stop = new Date(stop);
        }
    }

    /**
     * returns the duration in Milliseconds
     */
    duration() {
        if (this.stop) return this.stop.getTime() - this.start.getTime();
        else return new Date().getTime() - this.start.getTime();
    }

    /**
    * returns the hours as a decimal that the shift took to complete
    */
    getHours() {
        return this.duration() / 3600000 //MS in an Hour
    }

    getFormattedDuration(specificity = 2) {
        const duration = this.duration();
        const durationFormatted = timeFormat(duration, specificity);
        return durationFormatted;
    }

    equals(test, strict=true) {
        if(typeof test !== 'object') return false;
        else if (this.id && test.id && strict) {
            if(this.id === test.id) return true;
            else return false; 
        } else {
            if (this.category === test.category) {
                if (this.type === test.type) {
                    if (this.start.getTime() === test.start.getTime()) return true;
                    if (!strict) return true;
                }
            }
            return false;
        }
    }

    /**
     * sets the stop of the shift
     *
     * @param { Date } stop the stop Date/Time that this shift should stop, if none
     *          specified it defaults to now.
     */
    clockOut(stop = new Date()) {
        this.stop = stop;
    }

    /**
     * stops the currents shift at splitDateTime and starts a new shift 
     *      at the same moment of the specified type and category
     * 
     * @param { Date } splitDateTime the time to clock out the original shift
     * @param { String } type the type that the newshift should take on 
     * @param { String } category the category the newshift take on, the same category if none specified
     * @returns { Shift } returns the newly created shift 
     */
    splitShift(splitDateTime, type, category = this.category) {
        const newShift = new Shift({start: splitDateTime, type, category, stop: this.stop})
        this.clockOut(splitDateTime);
        return newShift;
    }

}


class ShiftCollection {
    static SORT_PARAMS = {
        START: "start",
        STOP: "stop",
        CATEGORY: "category",
        TYPE: "type",
        ID: "id",
        UID: "u_id",
        ASCENDING: "ascending",
        DESCENDING: "descending"
    };

    constructor(...newShifts) {
        this.shifts = [];
        this.add(...newShifts);
    }

    shiftsBy(sortParam = ShiftCollection.SORT_PARAMS.START, direction = ShiftCollection.SORT_PARAMS.DESCENDING) {
        const shifts = [...this.shifts];
        if(direction === ShiftCollection.SORT_PARAMS.ASCENDING)
            return shifts.sort((shift1, shift2) => {
                if(sortParam === ShiftCollection.SORT_PARAMS.STOP){
                    let stop1 = shift1.stop || new Date();
                    let stop2 = shift2.stop || new Date();
                    return stop1 > stop2 ? 1 : -1;
                }

                return shift1[sortParam] > shift2[sortParam] ? 1 : -1;
            });
        else {
            return shifts.sort((shift1, shift2) => {
                if (sortParam === ShiftCollection.SORT_PARAMS.STOP) {
                    let stop1 = shift1.stop || new Date();
                    let stop2 = shift2.stop || new Date();
                    return stop2 > stop1 ? 1 : -1;
                }
                return shift2[sortParam] > shift1[sortParam] ? 1 : -1;
            });
        }
    }

    toString() {
        return 'A ShiftCollection of length ' + this.length() + ' ' + this.shifts;
    }
    //future add with shifts being an object of objects of arrays
    // tempState = { ...state };
    // category = action.payload.category;
    // type = action.payload.type;
    // if(tempState[category]) {
    //     if (tempState[category][type]) {
    //         let temp = tempState[category][type];
    //         tempState[category][type] = [...temp, action.payload];
    //     } else {
    //         tempState[category][type] = [action.payload];
    //     }
    // } else {
    // tempState[category] = { [type]: [action.payload] };
    add(...newShifts) {
        for (let shift of newShifts) {
            if (!(shift instanceof Shift)) {
                shift = new Shift(shift);
            }
            this.shifts.push(shift);
        }
    }
    getTotalHours() {
        return this.shifts.reduce((totalHours, shift) => {
            return totalHours + shift.getHours();
        }, 0);
    }

    remove(removedShift) {
        const shiftIndex = this.shifts.findIndex((shift) => {
            if (shift.category === removedShift.category)
                if (shift.type === removedShift.type)
                    if (shift.start === removedShift.start) return true;
            return false;
        });

        if (shiftIndex !== -1) {
            this.shifts.splice(shiftIndex, 1);
            return true;
        }
        else return false;
    }

    contains(someShift, strict = false) {
        if (this.find(someShift, strict)) return true;
        return false;
    }
    /**
     * 
     * @param {string} initialCategory the initial category name to select the shifts
     * @param {string} newCategory the name all shifts selected should be changed to
     * @returns {boolean} true on success, false if no shifts of category initialCategory are found
     */
    changeCategory(initialCategory, newCategory) {
        const shifts = this.category(initialCategory).shifts; //shift collection is returned, get the array out of it

        if (shifts.length) {
            shifts.forEach(shift => { shift.category = newCategory });
            return true;
        }
        return false;
    }

    changeType(initialtype, newType) {
        const shifts = this.type(initialtype).shifts; //shift collection is returned, get the array out of it
        //if it's not empty
        if (shifts.length) {
            shifts.forEach(shift => { shift.type = newType });
            return true;
        }
        //emtpy, ogType doesn't exist
        return false;
    }

    find(someShift, strict = false) {
        for (let shift of this.shifts) {
            if(shift.equals(someShift, strict)) return shift;
        }
        return false;
    }

    getCategories() {
        const categories = new Set();
        for (let shift of this.shifts) {
            categories.add(shift.category);
        }
        return [...categories];
    }

    getTypes() {
        const types = new Set();
        for (let shift of this.shifts) {
            types.add(shift.type);
        }
        return [...types];
    }

    getShiftsAfter(dateTime) {
        return new ShiftCollection(...this.shifts
            .filter(({ start }) => dateTime.getTime() < start.getTime()));
    }
    getShiftsBefore(dateTime) {
        return new ShiftCollection(...this.shifts
            .filter(({ start }) => dateTime.getTime() < start.getTime()));
    }

    getCurrShifts() {
        return new ShiftCollection(...(this.shifts.filter((shift) => !shift.stop)));
    }

    category(queryString) {
        return this.filter('category', queryString);
    }

    type(queryString) {
        return this.filter('type', queryString);
    }

    filter(field, query) {
        const matched = this.shifts.filter(shift => shift[field] === query);
        return new ShiftCollection(...matched);
    }

    length() { return this.shifts.length };
}

/**
 * returns a date representing x days away from dateTime 
 * 
 * @param {number} days the number of days from dateTime you want
 * @param {Date} dateTime the dateTime you want to have a reference from(defaults to now)
 * @returns {Date} a new Date that is x hours from dateTime
 */
function daysFrom(days, dateTime = new Date()) {
    return new Date(dateTime.getTime() + (1000 * 60 * 60 * 24 * days)); //1000ms * 60s * 60m * 24hrs * X days
}

/**
 * returns a day representing x hours away from dateTime 
 * 
 * @param {number} hours the number of hours from dateTime you want
 * @param {Date} dateTime the dateTime you want to have a reference from(defaults to now)
 * @returns {Date} a new Date that is x hours from dateTime
 */
function hoursFrom(hours, dateTime = new Date()) {
    return new Date(dateTime.getTime() + (1000 * 60 * 60 * hours)); //1000ms * 60s * 60m * X hours
}

/**
 * returns a day representing x mins away from dateTime 
 * 
 * @param {number} mins the number of mins from dateTime you want
 * @param {Date} dateTime the dateTime you want to have a reference from(defaults to now)
 * @returns {Date} the Date that is x mins from dateTime
 */
function minsFrom(mins, dateTime = new Date()) {
    return new Date(dateTime.getTime() + (1000 * 60 * mins)); //1000ms * 60s * X mins
}

/**
 *
 * @param { number } ms A duration represented in Milliseconds negative values are returned as positive durations
 * @param { number } specificity How many scales of time you would like returned the biggest is always returned
 *
 */
function timeFormat(ms, specificity) {

    const MSINSEC = 1000;
    const MSINMINUTE = 60000;
    const MSINHOUR = 3600000;
    const MSINDAY = 86400000;
    const MSINMONTH = 2628000000;
    const MSINYEAR = MSINMONTH * 12;



    let result = '';
    let i = 0;
    if (ms < 0) {
        result += '- ';
        ms = ms * -1;
    }

    const y = Math.floor(ms / MSINYEAR);
    if (y) {
        i++;
        ms = ms - (y * MSINYEAR);
        result += y + ' years ';
        if (i === specificity) {
            return result;
        }
    }


    const mth = Math.floor(ms / MSINMONTH);
    if (mth) {
        i++;
        ms = ms - (mth * MSINMONTH);
        result += mth + ' Months ';
        if (i === specificity) {
            return result;
        }
    }

    const d = Math.floor(ms / MSINDAY);
    if (d) {
        i++;
        ms = ms - (d * MSINDAY);
        result += d + ' Days ';
        if (i === specificity) {
            return result;
        }
    }

    const h = Math.floor(ms / MSINHOUR);
    if (h) {
        i++;
        ms = ms - (h * MSINHOUR);
        result += h + ' Hours ';
        if (i === specificity) {
            return result;
        }
    }

    const min = Math.floor(ms / MSINMINUTE);
    if (min) {
        i++;
        ms = ms - (min * MSINMINUTE);
        result += min + ' mins ';
        if (i === specificity) {
            return result;
        }
    }
    const s = Math.floor(ms / MSINSEC);
    if (s) {
        i++;
        ms = ms - (s * MSINSEC);
        result += s + ' seconds ';
        if (i === specificity) {
            return result;
        }
    }
    /*
    if (ms) {
        i++;
        result += ms + ' ms ';
        if (i == specificity) {
            return result;
        }
    }
    */
    return result;

}
function timeFormatFromHours(hrs, specificity) {
    const ms = hrs * 3600000;
    return timeFormat(ms, specificity);
}

//testing purposes #cleanUp
// const deli = new ShiftCollection(...JSON.parse(localStorage.records));

export { ShiftCollection, Shift, minsFrom, hoursFrom, daysFrom, timeFormatFromHours }