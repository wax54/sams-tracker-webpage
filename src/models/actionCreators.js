import UserApi from "./Api";
import {v4 as uuid} from "uuid";
import UploadQueue from "./UploadQueue";
import { store } from "./store";
import { SET_TIMEFRAME } from "./actionTypes";

/** Timeframes */
export function setTimeFrame(timeframeKey) {
    return { type: SET_TIMEFRAME, payload: timeframeKey };
}

/** SHIFTS */
export function startShift(type, category) {
    const tempId = uuid();
    const newShift = {id: tempId, start: new Date(), type, category};
    
    // add a command and shift/data to the upload queue
    UploadQueue.addToQueue({type:"addShift", data: newShift});
    //add the shift to the store
    return { type: "START_SHIFT", payload: newShift };
}

export function resetShifts() {
    return { type: "RESET_SHIFTS" };
}

export function loadShifts() {
    return async function (dispatch) {
        const resp = await UserApi.getShifts();
        if (resp.status === true) {
            dispatch({ type: "ADD_SHIFTS", payload: resp.shifts });
            return true;
        }
        if (resp.status === false) {
            console.error(resp.errors);
            return false;
            //TODO make a queue for failed shift uploads to be merged later
            //dispatch({ type: "ADD_SHIFT_TO_UPLOAD_QUEUE", payload: newShift })
        }
    }
}

export function refreshShifts() {
    return async function (dispatch) {
        dispatch(resetShifts());
        return await dispatch(loadShifts());
    }
}

export function updateShift(shift) {
    return { type: "UPDATE_SHIFT", payload: shift };
}

export function updateAShift(shift) {
    UploadQueue.addToQueue({ type: "updateShift", data: shift });
    return updateShift(shift);
}

export function endShift(id) {
    const stop = new Date();
    return clockOutAt(id, stop);
}

export function clockOutAt(shiftId, stop) {
    const shift = store.getState().shifts[shiftId];
    shift.stop = stop;
    UploadQueue.addToQueue({type: 'updateShift', data: shift})
    return updateShift(shift);
}

export function deleteShift(shiftId) {
    UploadQueue.addToQueue({type: 'deleteShift', data: shiftId});
    return { type: "DELETE_SHIFT", shiftId: shiftId };
}

/** ERRORS (depreciated) */
export function resetErrors() {
    return { type: "RESET_ERRORS" };
}


/** Users */
export function setUser(userData) {
    return { type: "SET_USER", payload: userData };
}
export function resetUser() {
    return async function (dispatch) { 
        dispatch({ type: "RESET_USER" });
        dispatch(resetShifts());
        dispatch(resetGoals());
    }
}
export function authorizeUser({ username, password }) {
    return async function (dispatch) {
        const resp = await UserApi.login({username, password});
        if(resp.status === true)
            dispatch(setUser(resp.user));
            await dispatch(refreshShifts());
            await dispatch(refreshGoals());

        return resp;
    }
}

export function registerUser({ username, password }) {
    return async function (dispatch) {
        const resp = await UserApi.register({ username, password });
        if (resp.status === true)
            dispatch(setUser(resp.user));
            dispatch(resetShifts());
            dispatch(resetGoals());
        return resp;
    }
}

/** GOALS */
export function addGoal({type, category, seconds_per_day}) {
    const newGoal = { seconds_per_day, type, category };
    UploadQueue.addToQueue({type: "addGoal", data: newGoal});
    return {type: "ADD_GOAL", payload: newGoal};
}

export function resetGoals() {
    return { type: "RESET_GOALS" };
}

export function loadGoals() {
    return async function (dispatch) {
        const resp = await UserApi.getGoals();
        console.log(resp);
        if (resp.status === true) {
            dispatch({ type: "LOAD_GOALS", payload: resp.goals });
            return true;
        }
        if (resp.status === false) {
            console.error(resp.errors);
            return false;
        }
    }
}

export function refreshGoals() {
    return async function (dispatch) {
        dispatch(resetGoals());
        return await dispatch(loadGoals());
    }
}

export function updateGoal({type, category, seconds_per_day}) {
    return { type: "UPDATE_GOAL", payload: {type, category, seconds_per_day} };
}

export function updateAGoal(goal) {
    UploadQueue.addToQueue({type: "updateGoal", data: goal});
    return updateGoal(goal);
}

export function deleteGoal({ type, category }) {
    UploadQueue.addToQueue({type: "deleteGoal", data: { type, category }})
    return { type: "DELETE_GOAL", payload: { category, type } };
}