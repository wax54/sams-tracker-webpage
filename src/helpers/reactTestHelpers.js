import { render, waitFor } from '@testing-library/react';
import { MemoryRouter } from 'react-router';
import { Provider } from "react-redux";
import { store } from "../models/store"



function preRender(options = {}) {
    const defaultOptions = {
        initialEntries: ['/'],
    }
    const newOptions = {
        initialEntries:
            [...(options.initialEntries || []), ...defaultOptions.initialEntries],
        store: options.store || store
    }
    //console.log('options: ', options, 'newOptions', newOptions);
    return newOptions;
}

export function renderWithRedux(Component, options) {
    options = preRender(options);
    return render(
        <Provider store={options.store}>
            <MemoryRouter initialEntries={options.initialEntries}>
                <Component />
            </MemoryRouter>
        </Provider>
    );
}

export async function renderAndWaitFor(Component, options) {
    options = preRender(options);
    return await waitFor(() => render(
        <Provider store={options.store}>
            <MemoryRouter initialEntries={options.initialEntries}>
                <Component />
            </MemoryRouter>
        </Provider>
    ));
}
