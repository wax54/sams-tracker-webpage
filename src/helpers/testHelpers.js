import { render, waitFor } from '@testing-library/react';
import { MemoryRouter } from 'react-router';
import { Provider } from "react-redux";
import { store } from "../models/store"

function preRender(options = {}) {
    const defaultOptions = {
        initialEntries: ['/'],
    }
    const newOptions = {
        initialEntries:
            [...(options.initialEntries || []), ...defaultOptions.initialEntries],
        store: options.store || store
    }
    //console.log('options: ', options, 'newOptions', newOptions);
    return newOptions;
}

function renderWithRedux(Component, options) {
    options = preRender(options);
    return render(
        <Provider store={options.store}>
            <MemoryRouter initialEntries={options.initialEntries}>
                <Component />
            </MemoryRouter>
        </Provider>
    );
}

async function renderAndWaitFor(Component, options) {
    options = preRender(options);
    return await waitFor(() => render(
        <Provider store={options.store}>
            <MemoryRouter initialEntries={options.initialEntries}>
                <Component />
            </MemoryRouter>
        </Provider>
    ));
}



export {
    renderWithRedux,
    renderAndWaitFor
}