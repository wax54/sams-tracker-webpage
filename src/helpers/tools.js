import { ShiftCollection, Shift, daysFrom } from "../models/ShiftCollection";
export function makeColor (shift) {
    const str = shift.type + shift.category;
    let hash = 0;
    for (let i = 0; i < str.length; i++) {
        hash = str.charCodeAt(i) + ((hash << 5) - hash);
    }
    let colour = '#';
    for (let i = 0; i < 3; i++) {
        const value = (hash >> (i * 8)) & 0xFF;
        colour += ('00' + value.toString(16)).substr(-2);
    }
    return colour;
}

export function round(num, decimals = 0) {
    if (decimals < 0) {
        decimals = decimals * -1;
        const divider = 10 ** decimals;
        num = Math.round(num / divider) * divider;
    }
    const multiplier = 10 ** decimals;
    return Math.round(num * multiplier) / multiplier;
}

// shifts 
        // {
        //     _hours: 17.3,
        //     _categories: ["school", "my well being"],
        //     _currShifts:[shift],
        //     school: { 
        //         _hours: 12, 
        //         _types:["coding"], 
        //         coding: { 
        //             _hours: 12, 
        //             _shifts: [shift, shift, shift] 
        //         }
        //     },
        //     "my well being": {
        //         _hours: 5.3,
        //         _types: ["resting", "painting"],
        //         resting: {
        //             _hours: 2.7,
        //             _shifts: [shift,  shift]
        //         },
        //         painting: {
        //             _hours: 2.6,
        //             _shifts: [shift]
        //         }
        //     },
        //  }

export function getShiftsByCategory(storeShifts, timeFrame) {
    const shifts = {
        _allShifts: function () {
            const shifts = new ShiftCollection();
            this._categories.entries().forEach(cat => {
                this[cat]._types.forEach(type => {
                    this[cat][type]._shifts.forEach(shift => {
                        shifts.add(shift);
                    });
                });
            });
            return shifts;
        },
        _currShifts: [],
        _hours: 0,
        _categories: new Set()
    };
    const earliestDateInTimeFrame = daysFrom(-timeFrame.val);

    for (let key in storeShifts) {
        const shift = new Shift(storeShifts[key]);
        //only shows shifts in timeFrame
        if (shift.stop < earliestDateInTimeFrame) continue;

        const hours = shift.getHours();
        if (!shift.stop) shifts._currShifts.push(shift);
        //if there is no entry for this category...
        if (!shifts._categories.has(shift.category)) {
            //make it
            shifts._categories.add(shift.category);
            shifts[shift.category] = { _hours: 0, _types: new Set() };
        }
        //shifts[shift.category] is garunteed at this point
        //if there is no entry for this type in the category...
        if (!shifts[shift.category]._types.has(shift.type)) {
            // make it
            shifts[shift.category]._types.add(shift.type);
            shifts[shift.category][shift.type] = { _hours: 0, _shifts: [] };
        }
        shifts[shift.category][shift.type]._shifts.push(shift);
        shifts[shift.category][shift.type]._hours += hours;
        shifts[shift.category]._hours += hours;
        shifts._hours += hours;
    }
    return shifts;
}