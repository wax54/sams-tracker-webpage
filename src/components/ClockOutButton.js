import React from "react";
import { useDispatch, useSelector } from 'react-redux';
import { minsFrom } from '../models/ShiftCollection';
import { clockOutAt, deleteShift, endShift } from '../models/redux/actionCreators';
const NOW = "NOW";
const FIFTEEN_AGO = "-15";

const ClockOutButton = ({ shiftId }) => {
    const dispatch = useDispatch();
    const minsAgo = useSelector(({ shifts }) =>  (new Date() - new Date(shifts[shiftId].start)) / 1000 / 60); // 6000000ms / 1000 ms/s / 60 s/m = 100 minsAgo
    function remove() {
        dispatch(deleteShift(shiftId));
    }
    function clockOut(evt) {
        const clockoutDuration = evt.target.dataset.reference;

        if (clockoutDuration === NOW) {
            dispatch(endShift(shiftId));
        } else if (clockoutDuration === FIFTEEN_AGO) {
            const clockOut = minsFrom(0); //-15 minutes from now
            dispatch(clockOutAt(shiftId, clockOut));
        } else {
            console.error("ERROR ", clockoutDuration);
            dispatch(endShift(shiftId));
        }
    }
    return (
        <>
            <div className="dropdown clockout row align-items-start justify-content-center shadow border rounded">
                <button className="btn btn-primary col-10 text-center" 
                    style={{ borderTopRightRadius: 0, borderBottomRightRadius: 0 }}
                    data-reference={NOW} onClick={clockOut}
                >
                    Clock Out
                </button>
                <button className="btn btn-primary col-2 dropdown-toggle text-center" 
                    style={{ 
                        borderTopLeftRadius: 0, 
                        borderBottomLeftRadius: 0
                    }} 
                    data-bs-toggle="dropdown" 
                />
                
                <div className="dropdown-menu bg-primary clock-out p-0 shadow"
                    style={{
                        borderTopLeftRadius: 0,
                        borderTopRightRadius: 0
                    }} >
                    {minsAgo > 2 ? 
                        <button className="btn btn-block m-0 btn-primary text-light" data-reference={FIFTEEN_AGO} onClick={clockOut}>
                            15 MINUTES AGO
                        </button>
                        : null
                    }  

                    <button className="btn btn-block btn-danger text-light m-0" onClick={remove}>
                        CANCEL SHIFT
                    </button>
                </div>
            </div>
        </>
)};
export default ClockOutButton