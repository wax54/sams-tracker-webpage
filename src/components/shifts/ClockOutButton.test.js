import { renderWithRedux, renderAndWaitFor } from '../../helpers/reactTestHelpers';
import ClockOutButton from './ClockOutButton';
import { createStore } from 'redux';
import userEvent from '@testing-library/user-event';
import { store } from '../../models/store';
import { resetAll, startShift } from '../../models/actionCreators';
import { minsFrom, Shift } from '../../models/ShiftCollection';


const start = new Date();
const stop = minsFrom(10, start)
const TEST_SHIFT = new Shift({ id: 0, type: "testtype", category: "testcategory", start, stop });
const addShift = (shift) => ({ type: "START_SHIFT", payload: shift });


beforeEach(() => {
    store.dispatch(resetAll());
    store.dispatch(addShift(TEST_SHIFT));
});

test("doesn't blow up", () => {
    renderWithRedux(() => <ClockOutButton shiftId={TEST_SHIFT.id} />);
    renderWithRedux(() => <ClockOutButton shiftId={'non existent'} />);
});

test("matchesSnapshot", () => {
    const { asFragment:withValidShiftId} = renderWithRedux(() => 
        <ClockOutButton shiftId={TEST_SHIFT.id} />
    );

    expect(withValidShiftId()).toMatchSnapshot();

    const { asFragment: withoutValidShiftId } = renderWithRedux(() => 
        <ClockOutButton shiftId={'non existent'} />
    );
    expect(withoutValidShiftId()).toMatchSnapshot();
    
});


test("Default Buttons on Page", () => {
    const { getByText, } = renderWithRedux(() => <ClockOutButton shiftId={TEST_SHIFT.id} />);
    expect(getByText("Clock Out")).toBeInTheDocument();
    expect(getByText("CANCEL SHIFT")).toBeInTheDocument();
});

test("clockOut 15mins ago appears when shift older than 15mins", () => {
    
    const olderThan15Mins = {
        ...TEST_SHIFT,
        id: 'testShift2',
        start: minsFrom(-16),
    }

    store.dispatch(addShift(olderThan15Mins));

    const { getByText: getOlderByText } = renderWithRedux(() => <ClockOutButton shiftId={olderThan15Mins.id} />);
    expect(getOlderByText("15 MINUTES AGO")).toBeInTheDocument();
});

test("clockOut 15mins ago Does not appears when shift older than 15mins", () => {

    const youngerThan15Mins = {
        ...TEST_SHIFT,
        id: 'testShift1',
        start: minsFrom(-14)
    }

    store.dispatch(addShift(youngerThan15Mins));

    const { queryByText: queryYoungerByText } = renderWithRedux(() => <ClockOutButton shiftId={youngerThan15Mins.id} />);
    expect(queryYoungerByText("15 MINUTES AGO")).not.toBeInTheDocument();
});


test("deletes shift when CANCEL SHIFT is pressed", () => {
    const { getByText } = renderWithRedux(() => 
        <ClockOutButton shiftId={TEST_SHIFT.id} />
    );
    const cancelBtn = getByText("CANCEL SHIFT");

    expect(Object.keys(store.getState().shifts).length).toBe(1);
    userEvent.click(cancelBtn);
    expect(Object.keys(store.getState().shifts).length).toBe(0);

});