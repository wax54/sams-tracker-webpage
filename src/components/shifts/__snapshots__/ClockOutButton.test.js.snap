// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matchesSnapshot 1`] = `
<DocumentFragment>
  <div
    class="dropdown clockout row align-items-start justify-content-center shadow border rounded"
  >
    <button
      class="btn btn-primary col-10 text-center"
      data-reference="NOW"
      style="border-top-right-radius: 0; border-bottom-right-radius: 0;"
    >
      Clock Out
    </button>
    <button
      class="btn btn-primary col-2 dropdown-toggle text-center"
      data-bs-toggle="dropdown"
      style="border-top-left-radius: 0; border-bottom-left-radius: 0;"
    />
    <div
      class="dropdown-menu bg-primary clock-out p-0 shadow"
      style="border-top-left-radius: 0; border-top-right-radius: 0;"
    >
      <button
        class="btn btn-block btn-danger text-light m-0"
      >
        CANCEL SHIFT
      </button>
    </div>
  </div>
</DocumentFragment>
`;

exports[`matchesSnapshot 2`] = `<DocumentFragment />`;
